virtcontainers/persist.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2018 Huawei Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"errors"
+
+	"github.com/kata-containers/runtime/virtcontainers/device/api"
+	exp "github.com/kata-containers/runtime/virtcontainers/experimental"
+	"github.com/kata-containers/runtime/virtcontainers/persist"
+	persistapi "github.com/kata-containers/runtime/virtcontainers/persist/api"
+	"github.com/kata-containers/runtime/virtcontainers/types"
+)
+
+var (
+	errContainerPersistNotExist = errors.New("container doesn't exist in persist data")
+)
+
+func (s *Sandbox) dumpState(ss *persistapi.SandboxState, cs map[string]persistapi.ContainerState) error {
+	ss.SandboxContainer = s.id
+	ss.GuestMemoryBlockSizeMB = s.state.GuestMemoryBlockSizeMB
+	ss.State = string(s.state.State)
+
+	for id, cont := range s.containers {
+		state := persistapi.ContainerState{}
+		if v, ok := cs[id]; ok {
+			state = v
+		}
+		state.State = string(cont.state.State)
+		state.Rootfs = persistapi.RootfsState{
+			BlockDeviceID: cont.state.BlockDeviceID,
+			FsType:        cont.state.Fstype,
+		}
+
+		// FIXME: container should store its own PID
+		if id == s.id {
+			state.ShimPid = s.state.Pid
+		} else {
+			state.ShimPid = cont.state.Pid
+		}
+		cs[id] = state
+	}
+
+	// delete removed containers
+	for id := range cs {
+		if _, ok := s.containers[id]; !ok {
+			delete(cs, id)
+		}
+	}
+
+	return nil
```

Comment:
It looks like you could drop the `error` return here as it isn't really needed.

Reply:
Do you mean we should modify 
```
func (s *Sandbox) dumpState(ss *persistapi.SandboxState, cs map[string]persistapi.ContainerState) error
```

to 
```
func (s *Sandbox) dumpState(ss *persistapi.SandboxState, cs map[string]persistapi.ContainerState) 
```

?

Since this is a `SetFunc`, I think it's better to keep the error in case we need it in future. Though we don't have an error returned yet.